include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h

```diff
@@ -188,20 +190,39 @@ struct __parallel_merge_submitter<_IdType, __internal::__optional_kernel_name<_N
 
         const _IdType __steps = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __chunk);
 
-        auto __event = __exec.queue().submit(
-            [&__rng1, &__rng2, &__rng3, __comp, __chunk, __steps, __n1, __n2](sycl::handler& __cgh) {
-                oneapi::dpl::__ranges::__require_access(__cgh, __rng1, __rng2, __rng3);
-                __cgh.parallel_for<_Name...>(sycl::range</*dim=*/1>(__steps), [=](sycl::item</*dim=*/1> __item_id) {
-                    const _IdType __i_elem = __item_id.get_linear_id() * __chunk;
-                    const auto __start =
-                        __find_start_point(__rng1, _IdType{0}, __n1, __rng2, _IdType{0}, __n2, __i_elem, __comp);
-                    __serial_merge(__rng1, __rng2, __rng3, __start.first, __start.second, __i_elem, __chunk, __n1, __n2,
-                                   __comp);
-                });
+        using __val_t = _split_point_t<_IdType>;
+        using __result_and_scratch_storage_t = __result_and_scratch_storage<_ExecutionPolicy, __val_t>;
+        auto __p_res_storage = new __result_and_scratch_storage_t(__exec, 1, 0);
+
+        // Save the raw pointer into a shared_ptr to return it in __future and extend the lifetime of the storage.
+        std::shared_ptr<__result_and_scratch_storage_base> __p_result_base(__p_res_storage);
+
+        auto __event = __exec.queue().submit([&__rng1, &__rng2, &__rng3, __p_res_storage, __comp, __chunk, __steps, __n,
+                                              __n1, __n2](sycl::handler& __cgh) {
+            oneapi::dpl::__ranges::__require_access(__cgh, __rng1, __rng2, __rng3);
+            auto __result_acc =
+                __p_res_storage->template __get_result_acc<sycl::access_mode::write>(__cgh, __dpl_sycl::__no_init{});
+
+            __cgh.parallel_for<_Name...>(sycl::range</*dim=*/1>(__steps), [=](sycl::item</*dim=*/1> __item_id) {
+                auto __id = __item_id.get_linear_id();
+                const _IdType __i_elem = __id * __chunk;
+
+                const auto __n_merge = std::min<_IdType>(__chunk, __n - __i_elem);
+                const auto __start =
+                    __find_start_point(__rng1, _IdType{0}, __n1, __rng2, _IdType{0}, __n2, __i_elem, __comp);
+                auto __ends = __serial_merge(__rng1, __rng2, __rng3, __start.first, __start.second, __i_elem, __n_merge,
+                                             __n1, __n2, __comp, __n);
+
+                if (__id == __steps - 1) //the last WI does additional work
```

Comment:
This branch can probably be combined with the one in `std::min` above, where we just have a full different implementation for the "final" work-item.  
Sadly, because `__chunk` is a runtime param based on device type, we probably can't unroll the serial merge loop either way, but this still could offer some benefit.

Reply:
I'm afraid I didn't catch what concrete code you suggest....

Reply:
> but opportunity for performance gains

Why?

Reply:
Its optional, it may not have any measurable benefit, just a suggestion.

Reply:
Why  "this reduces the number of them"?
We have one if_else in my variant and we have one if_else in your variant (plus code duplication -  the which is executed within "if" and within "else").

Reply:
`min` includes an `if`.  In this case, the branching within `min` exactly will coincide with the explicit `if` you have.
The PR has both a `min`, and an `if` / `else`.
My suggestion just has a `if`/ `else` with no `min`.  

Its likely quite minor, but I doubt the compiler can group these on its own.

Reply:
Oh...  Do you mean `std::min<_IdType>(__chunk, __n - __i_elem);` also produces a branch?
Probably...But I'm not sure. I have a guess that there is "min" as one low level instruction... no any branches in kernel.. No?

Reply:
Me calling it a "branch" is probably a simplification.
The compiler can probably do better than merely an `if` statement, but certainly it is more expensive than a simple assignment (as in my suggestion).
Anyway, this is a minor thing and we can drop it in favor of the other stuff that should have higher priority.

Reply:
I mean std::min is compiled into just one instruction to  `minsd`, like an assignment is compiled also in one instruction  `mov`. 
No?

Reply:
All instructions are not created equal from a timing perspective, but again, lets drop this discussion for now.